### ISSUE TITLE
the size of the iterator changes after the first epoch

### ISSUE THREAD

**Issue by twmht** [NVIDIA/DALI#4440]
Hello,

I found out the size of dali iterator changes after the first epoch,

for example,  I defined Dali dataloader as the following,

```python
class DaliLoader():
    def __init__(...):
        self._dali_iterator = DALIGenericIterator([pipe], ['img', 'gt_label'], reader_name='Reader')
    def __len__(self):
       # this is how I define the length of my dataloader 
        return int(math.ceil(self._dali_iterator._size / self._dali_iterator.batch_size))
    def __next__(self):
            try:
                data = next(self._dali_iterator)
            except StopIteration:
                self._dali_iterator.reset()
                raise StopIteration
        return data
```

and I use the normal pytorch training flow to load the data

```python
for idx, data in enumerate(data_loader):
    pass
```

the first epoch of `self._dali_iterator._size` is 2848800 in each gpu, the length of the dataloader is 14244

but after the first epoch finished, the size is changed to 2848700, and the length of the dataloader is 14243.

there is a missing batch.

Here are the detail,

I use 6 gpus for training ddp, the total number of data is 17091657, all the data are stored in the mxnet recordIO.

the pipeline is very similar to https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/dataset.py#L216

any idea?

**Comment by JanuszL** [comment]
Hi @twmht,

This can happen and is somehow expected. Let's consider a dataset with 10 samples, no shuffling, and batch size 10. DALI makes sure that each epoch of each sample in the data set is seen at least once. During the first epoch, the last batch wraps the data set up and consists of [9, 0, 1] samples. The next epoch is shorter and has only 3 iterations.
If you want to make each epoch the same length, set `pad_last_batch=True` in the reader and `last_batch_policy` in the iterator accordingly.

**Comment by twmht** [comment]
set `pad_last_batch=True` works. thank you @JanuszL
